src/wallet/rpcwallet.cpp

```diff
@@ -598,6 +653,103 @@ UniValue listaddresses(const UniValue& params, bool fHelp)
         entry.pushKV("source", "legacy_hdseed");
 
         bool hasData = add_sapling(saplingAddresses, PaymentAddressSource::LegacyHDSeed, entry);
+        if (hasData) {
+            ret.push_back(entry);
+        }
+    }
+
+    // mnemonic seed source
+    {
+        UniValue entry(UniValue::VOBJ);
+        entry.pushKV("source", "mnemonic_seed");
+
+        bool hasData = false;
+        // transparent
+        // For each address in pwalletMain->mapKeyMetadata, include any address
+        // for which its HD keypath metadata indicates it is associated with the legacy
+        // account ID. All such addresses are derived from the mnemonic seed.
+        UniValue mnemonic_taddrs(UniValue::VARR);
+        for (const auto& [keyId, keyMeta] : pwalletMain->mapKeyMetadata) {
+            auto account = libzcash::ParseHDKeypathAccount(44, Params().BIP44CoinType(), keyMeta.hdKeypath);
+            if (account.has_value() && account.value() == ZCASH_LEGACY_ACCOUNT) {
+                mnemonic_taddrs.push_back(keyIO.EncodeDestination(keyId));
+                hasData = true;
+            }
+        }
```

Comment:
Instead of doing this iteration again, we should collect the mnemonic-associated t-addrs at the same time we are already iterating over other t-addrs. Then we can use `PaymentAddressSource` to distinguish them, which should already be implementing this particular distinction for us (and if it isn't correctly, we should ensure it does).

Reply:
i.e. this would become:
```suggestion
        if (!t_mnemonic_dests.empty()) {
            UniValue mnemonic_t_addrs(UniValue::VARR);
            for (const CTxDestination& dest : t_mnemonic_dests) {
                mnemonic_t_addrs.push_back(keyIO.EncodeDestination(dest));
            }
            mnemonic_transparent.pushKV("addresses", random_t_addrs);
            hasData = true;
        }
```